### PR TITLE
Added Encoder and Decoder Labels

### LINKED ISSUE #989: Encoder/Decoder labels
The Encoder/Decoder components are both missing labels on their ports

We would like to have something similar to the Mux's

Ideally, it would be:
Inputs are labeled: `I0` - `I##`
and outputs are labeled: `O0` - `O##`

Encoder/Decoder currently:
Screen Shot 2022-02-11 at 2 18 46 PM

Muxs currently:
Screen Shot 2022-02-11 at 2 18 45 PM
.

### PR DESCRIPTION
**Pull request by mattsweet417**
Encoders and Decoders do not have any labels on their ports, so I added an updatePortNames method to both the Encoder and Decoder that is very similar to Mux.

Fixes #989

### CODE REVIEW

**Review by TGCrystal — CHANGES_REQUESTED** [review]
When the size of the encode/decoder is increased, labels are not applied to the additional inputs/outputs. This can probably be fixed by calling that `updatePortNames` function when `CoderPortChangeAction` is executed/undone.

**Comment by mattsweet417** [comment]
Should be fixed now, sorry about that!

File: package.json [hunk 1]
```diff
@@ -75,5 +75,6 @@
       "last 1 firefox version",
       "last 1 safari version"
     ]
-  }
+  },
+  "dependencies": {}
```
Comment (TGCrystal) [inline package.json:79]: This was probably accidentally caught up in a commit, I don't think it seems like it should impact the functionality of the pr. Please revert this change.

**Review by TGCrystal — CHANGES_REQUESTED** [review]
Looks good now, although there is one issue where I'm not sure what the right approach to fix it is. When the encoder/decoder is increased to a large size, the text ends up overlapping (see below).
image
@LeonMontealegre What is the best approach to deal with this? Changing the text size? Changing the encoder width?

**Comment by LeonMontealegre** [comment]
I believe the `Mux` increase in width slightly as it increases in # of ports, so probably doing that
@TGCrystal @mattsweet417

**Comment by mattsweet417** [comment]
The inputs/outputs shouldn't overlap now, I added a multiplier to the width of the encoder/decoder when inputs go over 5.

**Review by TGCrystal — DISMISSED** [review]

File: src/app/digital/actions/ports/CoderPortChangeAction.ts [hunk 2]
```diff
@@ -53,7 +53,10 @@ export class CoderPortChangeAction implements Action {
      * @param val refers to the target number chosen by the user.
      */
     protected changeSize(val: number): void {
-        this.obj.setSize(V(DEFAULT_SIZE, DEFAULT_SIZE/2 * Math.pow(2, val)));
+        let mul: number = 1;
+        if(val >= 5) mul = 1.2;
+        if(val == 8) mul = 1.4;
```
Comment (LeonMontealegre) [inline src/app/digital/actions/ports/CoderPortChangeAction.ts:58]: Should probably just have this be a continuous function instead of having it change every 5,8 steps
So like just `DEFAULT_SIZE * (1 + (val - 1)/20)` or something that works

File: src/app/digital/models/ioobjects/other/Decoder.ts [hunk 3]
```diff
@@ -36,6 +38,15 @@ export class Decoder extends DigitalComponent {
         });
     }
 
+    public updatePortNames(): void {
+        this.inputs.getPorts().forEach((p, i) => {
+            if (p.getName() == "") p.setName(`I${i}`);
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Decoder.ts:43]: nit: use `===`

File: src/app/digital/models/ioobjects/other/Decoder.ts [hunk 4]
```diff
@@ -36,6 +38,15 @@ export class Decoder extends DigitalComponent {
...
+        })
+        this.outputs.getPorts().forEach((p, i) => {
+            if (p.getName() == "") p.setName(`O${i}`);
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Decoder.ts:46]: nit: use `===`

File: src/app/digital/models/ioobjects/other/Decoder.ts [hunk 5]
```diff
@@ -36,6 +38,15 @@ export class Decoder extends DigitalComponent {
...
+        })
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Decoder.ts:47]: nit: semicolon

File: src/app/digital/models/ioobjects/other/Decoder.ts [hunk 6]
```diff
@@ -36,6 +38,15 @@ export class Decoder extends DigitalComponent {
         });
     }
 
+    public updatePortNames(): void {
+        this.inputs.getPorts().forEach((p, i) => {
+            if (p.getName() == "") p.setName(`I${i}`);
+        })
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Decoder.ts:44]: nit: semicolon

File: src/app/digital/models/ioobjects/other/Encoder.ts [hunk 7]
```diff
@@ -38,6 +39,15 @@ export class Encoder extends DigitalComponent {
         });
     }
 
+    public updatePortNames(): void {
+        this.inputs.getPorts().forEach((p, i) => {
+            if (p.getName() == "") p.setName(`I${i}`);
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Encoder.ts:44]: nit: use `===`

File: src/app/digital/models/ioobjects/other/Encoder.ts [hunk 8]
```diff
@@ -38,6 +39,15 @@ export class Encoder extends DigitalComponent {
...
+        })
+        this.outputs.getPorts().forEach((p, i) => {
+            if (p.getName() == "") p.setName(`O${i}`);
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Encoder.ts:47]: nit: use `===`

File: src/app/digital/models/ioobjects/other/Encoder.ts [hunk 9]
```diff
@@ -38,6 +39,15 @@ export class Encoder extends DigitalComponent {
...
+        })
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Encoder.ts:48]: nit: semicolon

File: src/app/digital/models/ioobjects/other/Encoder.ts [hunk 10]
```diff
@@ -38,6 +39,15 @@ export class Encoder extends DigitalComponent {
         });
     }
 
+    public updatePortNames(): void {
+        this.inputs.getPorts().forEach((p, i) => {
+            if (p.getName() == "") p.setName(`I${i}`);
+        })
```
Comment (LeonMontealegre) [inline src/app/digital/models/ioobjects/other/Encoder.ts:45]: nit: semicolon

**Review by LeonMontealegre — CHANGES_REQUESTED** [review]
Looking good! just need a few more changes!

**Review by LeonMontealegre — APPROVED** [review]

**Review by TGCrystal — APPROVED** [review]
